### PR TITLE
Add restart data output for Network option data (keywords NETWORK, BRANPROP NODEPROP)

### PR DESCRIPTION
**Pull request by jalvestad**
This pull request implements code for output of the so-called extended Network option in Eclipse using the keywords BRANPROP and NODEPROP in addition to the NETORK keyword in the Runspec section. 

The restart data consists of the tables:
INODE
IBRAN
INOBR
ZNODE
RNODE 
RBRAN

and the content of the tables are documented in the code. 

Acknowledgement to Torbjørn Skille for identifying the meaning/content of the restart-data tables. 

The pull-request is ready for review and merge after modifications based comments from reviewers from my side. 

The pull-request pass the normal ctest - tests in opm-common. However, several elements of the INTEHEAD vector is modified, hence many of the opm-simulator tests fails, and update data is needed since the content of the restart file is changed.

### CODE REVIEW

**Comment by joakim-hove** [comment]
jenkins build this please

File: src/opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.cpp [hunk 1]
```diff
@@ -147,6 +158,26 @@ void ExtNetwork::add_node(Node node)
...
+{
+    this->insert_indexed_node_names.emplace_back(name);
+}
+
+bool ExtNetwork::has_indexed_node_name(std::string name) const
+{
+    // Find given element in vector
+    auto it = std::find(this->insert_indexed_node_names.begin(), this->insert_indexed_node_names.end(), name);
+
+    if (it != this->insert_indexed_node_names.end())
```
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.cpp:171]: If you want to save a couple of bytes of source code this could be:
```
return (it != this->insert_indexed_node_names.end());
```

File: opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.hpp [hunk 2]
```diff
@@ -38,14 +38,19 @@ class ExtNetwork {
     ExtNetwork() = default;
     bool active() const;
     bool has_node(const std::string& name) const;
+    bool has_indexed_node_name(std::string name) const;
```
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.hpp:41]: The functions `has_indexed_node_name()` and `add_indexed_node_name()` should be private, and they should be called internally from the `add_branch()` method.

File: opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.hpp [hunk 3]
```diff
@@ -38,14 +38,19 @@ class ExtNetwork {
...
+    void add_indexed_node_name(std::string name);
     void add_branch(Branch branch);
     void drop_branch(const std::string& uptree_node, const std::string& downtree_node);
     const Node& node(const std::string& name) const;
     const Node& root() const;
     std::vector<Branch> downtree_branches(const std::string& node) const;
+    std::vector<const Branch*> branches() const;
     std::optional<Branch> uptree_branch(const std::string& node) const;
     std::vector<std::string> node_names() const;
+    std::vector<std::string> insert_index_nd_names() const;
```
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.hpp:52]: Is it necessary to add a new `insert_index_nd_names()` method - I would guess the existing `node_names()` function could be changed to return the new `insert_indexed_node_names()` vector?

File: tests/test_AggregateActionxData.cpp [hunk 4]
```diff
@@ -128,7 +128,7 @@ BOOST_AUTO_TEST_CASE (Declared_Actionx_data)
     double secs_elapsed = 3.1536E07;
     const auto ih = Opm::RestartIO::Helpers::
         createInteHead(es, grid, sched, secs_elapsed,
-                       rptStep, rptStep, rptStep);
```
Comment (joakim-hove) [inline tests/test_AggregateActionxData.cpp:131]: This was a totally unrelated change - might induce a data update and should be in a separate PR.

File: tests/test_AggregateNetworkData.cpp [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+
+    start = 4*ih[VI::NIBRAN];
+    BOOST_CHECK_EQUAL(iBran[start + 0], 6);
+    BOOST_CHECK_EQUAL(iBran[start + 1], 3);
+    BOOST_CHECK_EQUAL(iBran[start + 2], 8);
+
+    //ZNode-parameters
+    const auto& zNode = networkData.getZNode();
+    start = 0*ih[VI::NZNODE];
+    BOOST_CHECK_EQUAL(zNode[start + 0].c_str(), "B1      ");
```
Comment (joakim-hove) [inline tests/test_AggregateNetworkData.cpp:217]: Try to reduce the dependency on counting exactly eight spaces; something like:

```C++
std::string pad8(const std::string& s) {
     return s + std::string( 8 - s.size(), ' ');   
}

const std::blank8 = "        ";
.....
BOOST_CHECK_EQUAL(zNode[start + 0], pad8("N1"));
BOOST_CHECK_EQUAL(zNode[start + 1], blank8);
```

**Comment by joakim-hove** [comment]
OK - this looks good. A general comment: in addition to the main changes related to restart output of network++ properties there are some unrelated changes - at least:

1. Some changes in the network api - with the `indexed_names` management.
2. `Schedule::hasGroup()`
3. A +1 change in `test_AggregateActionxData`
4. New member `NetworkDims` in `Runspec`

I think small isolated issues like these should be submitted as isolated PR's while work still continues on the main PR. The advantages of that is:

1. Small PR's - easy and fast to review and merge.
2. Better review possibilities.
3. Give people a heads up of what is coming.
4. It gives a nice and warm feeling of progress to the small PR's trickle in!

Not necessary to extract separate PR's now; personally I would do it because I actually think it would speed up the process - but your YMMV.

File: src/opm/output/eclipse/AggregateNetworkData.cpp [hunk 6]
```diff
@@ -0,0 +1,752 @@
...
+        if (branch   == nullptr) {
+            continue;
+        }
+
+        branchOp(*branch, branchID - 1);
+    }
+}
+
+template < typename T>
+std::pair<bool, int > findInVector(const std::vector<T>  & vecOfElements, const T  & element)
```
Comment (joakim-hove) [inline src/opm/output/eclipse/AggregateNetworkData.cpp:113]: Funny white space in the signature?

File: src/opm/output/eclipse/AggregateNetworkData.cpp [hunk 7]
```diff
@@ -0,0 +1,752 @@
...
+{
+    std::pair<bool, int > result;
```
Comment (joakim-hove) [inline src/opm/output/eclipse/AggregateNetworkData.cpp:115]: No - this is what we have `std::optional<int>` for.

File: src/opm/output/eclipse/AggregateNetworkData.cpp [hunk 8]
```diff
@@ -0,0 +1,752 @@
...
+    }
+    else
+    {
+        result.first = false;
+        result.second = -1;
+    }
+    return result;
+}
+
+int next_branch(int node_no, std::vector<int> inlets, std::vector<int> outlets)
```
Comment (joakim-hove) [inline src/opm/output/eclipse/AggregateNetworkData.cpp:133]: I guess the `inlets` and `outlets`argument could be `const std::vector<int>&`?

File: src/opm/output/eclipse/AggregateNetworkData.cpp [hunk 9]
```diff
@@ -0,0 +1,752 @@
...
+    // for nodes that are not well groups, set the node pressure to the fixed pressure potentially higher in the node tree
+    if (!node_wgroup) {
+        if (fixedPressureNode(sched, nodeName, lookup_step)) {
+            // node is a fixed pressure node
+            node_pres = units.from_si(M::pressure, network.node(nodeName).terminal_pressure().value());
+        }
+        else {
+            // find fixed pressure higher in the node tree
+            bool fp_flag = false;
+            auto node_name = nodeName;
```
Comment (joakim-hove) [inline src/opm/output/eclipse/AggregateNetworkData.cpp:246]: What is the point of variable `node_name = nodeName`?

File: src/opm/output/eclipse/AggregateNetworkData.cpp [hunk 10]
```diff
@@ -0,0 +1,752 @@
...
+                if (fixedPressureNode(sched, upt_br.uptree_node(), lookup_step)) {
+                    node_pres = units.from_si(M::pressure, network.node(upt_br.uptree_node()).terminal_pressure().value());
+                    fp_flag = true;
+                } else {
+                    node_name = upt_br.uptree_node();
+                    if (network.uptree_branch(node_name).has_value()) {
+                        upt_br = network.uptree_branch(node_name).value();
+                    } else {
+                        std::stringstream str;
+                        str << "Node: " << nodeName << " has no uptree node with fixed pressure condition, uppermost node:  " << node_name;
```
Comment (joakim-hove) [inline src/opm/output/eclipse/AggregateNetworkData.cpp:258]:
```C++
#include <fmt/format.h>
....
msg = fmt::format("Node: {0} has no uptree ....  node: {0}", node_name);
```

File: src/opm/output/eclipse/AggregateNetworkData.cpp [hunk 11]
```diff
@@ -0,0 +1,752 @@
...
+    double t_wpr = 0.;
+    double t_gpr = 0.;
+
+    const auto& grp = sched.getGroup(groupName, lookup_step);
+    // Calculate average flow rate and surface condition densities for wells in group
+    for (const auto& well_name : grp.wells()) {
+        const auto& well = sched.getWell(well_name, lookup_step);
+        if (well.isProducer()) {
+            const auto& pvtNum = well.pvt_table_number();
+            t_opr = get(well.name(),"WOPR");
```
Comment (joakim-hove) [inline src/opm/output/eclipse/AggregateNetworkData.cpp:310]: This should use: `smry.get_well_var(well.name(), "WOPR", 0.0);`

File: src/opm/output/eclipse/AggregateNetworkData.cpp [hunk 12]
```diff
@@ -0,0 +1,752 @@
...
+int numberOfBranchesConnToNode(const Opm::Schedule& sched, const std::string& nodeName, const size_t lookup_step)
+{
+    auto& network = sched[lookup_step].network();
+    if (network.has_node(nodeName)) {
+        int noBranches = network.downtree_branches(nodeName).size();
+        noBranches = (network.uptree_branch(nodeName).has_value()) ? noBranches+1 : noBranches;
+        return noBranches;
+    } else {
+        std::stringstream str;
+        str << "actual node: " << nodeName << " has not been defined at report time: " << lookup_step+1;
```
Comment (joakim-hove) [inline src/opm/output/eclipse/AggregateNetworkData.cpp:430]: Use `fmt::format()`

File: src/opm/output/eclipse/AggregateNetworkData.cpp [hunk 13]
```diff
@@ -0,0 +1,752 @@
...
+{
+    //
+    using Ix = VI::INode::index;
+    iNode[Ix::NoBranchesConnToNode] = numberOfBranchesConnToNode(sched, nodeName, lookup_step);
+    iNode[Ix::CumNoBranchesConnToNode] = cumNumberOfBranchesConnToNode(sched, nodeName, lookup_step);
+    if (sched.hasGroup(nodeName, lookup_step)) {
+        iNode[Ix::Group] = sched.getGroup(nodeName, lookup_step).insert_index();
+    }
+    iNode[Ix::FixedPresNode] = (fixedPressureNode(sched, nodeName, lookup_step)) ? 1 : 0;
+    iNode[4] = 1;
```
Comment (joakim-hove) [inline src/opm/output/eclipse/AggregateNetworkData.cpp:472]: What is `4`?

File: src/opm/output/eclipse/AggregateNetworkData.cpp [hunk 14]
```diff
@@ -0,0 +1,752 @@
...
+                    const Opm::SummaryState&  sumState,
+                    const std::string&        nodeName,
+                    const std::size_t         lookup_step,
+                    const Opm::UnitSystem&    units,
+                    RNodeArray&               rNode)
+{
+    using Ix = ::Opm::RestartIO::Helpers::VectorItems::RNode::index;
+    // node dynamic pressure
+    std::string compKey = "GPR:" + nodeName;
+    rNode[Ix::NodePres] = (sumState.has(compKey)) ? sumState.get(compKey) : 0.;
```
Comment (joakim-hove) [inline src/opm/output/eclipse/AggregateNetworkData.cpp:595]: `rNode[Ix::NodePres] = sumState.get_group_var(nodeName, "GPR", 0);`

File: src/opm/output/eclipse/AggregateNetworkData.cpp [hunk 15]
```diff
@@ -0,0 +1,752 @@
...
+
+    // equal to 0. for fixed pressure nodes, 1. otherwise
+    rNode[Ix::FixedPresNode] = (fixedPressureNode(sched, nodeName, lookup_step)) ? 0. : 1.;
+
+    // equal to i) highest well p_thp if wellgroup and ii) pressure of uptree node with fixed pressure
+    rNode[Ix::PressureLimit] = nodePressure(sched, sumState, nodeName, units, lookup_step);
+
+    // fixed value
+    rNode[15] = 1.;
```
Comment (joakim-hove) [inline src/opm/output/eclipse/AggregateNetworkData.cpp:604]: 15?

File: src/opm/output/eclipse/AggregateNetworkData.cpp [hunk 16]
```diff
@@ -0,0 +1,752 @@
...
+        IBran::staticContrib(sched, branch, lookup_step, ib);
+    });
+
+    // Define Static Contributions to INobr Array
+    const std::vector<int> inobr = inobrFunc(sched, lookup_step);
+
+    // Define Static Contributions to INobr Array
+    if (inobr.size() != entriesPerInobr(inteHead)) {
+        std::stringstream str;
+        str << "Actual size of inobr:  " << inobr.size() << " different from required size: " << entriesPerInobr(inteHead);
```
Comment (joakim-hove) [inline src/opm/output/eclipse/AggregateNetworkData.cpp:718]: `fmt::format()`

File: src/opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.cpp (hunk 1 above)
Comment (jalvestad) [reply src/opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.cpp:171]: OK - done!

File: opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.hpp (hunk 2 above)
Comment (jalvestad) [reply opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.hpp:41]: OK - fine!

File: opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.hpp (hunk 3 above)
Comment (jalvestad) [reply opm/parser/eclipse/EclipseState/Schedule/Network/ExtNetwork.hpp:52]: OK - I have modified the node_names() function.

File: tests/test_AggregateActionxData.cpp (hunk 4 above)
Comment (jalvestad) [reply tests/test_AggregateActionxData.cpp:131]: This is a code correction that will not trigger a change for the actual test, and has no further implications. I can remove it but cannot promise a separate pull request for it.

Comment (joakim-hove) [reply tests/test_AggregateActionxData.cpp:131]: ok - just leave it

File: tests/test_AggregateNetworkData.cpp (hunk 5 above)
Comment (jalvestad) [reply tests/test_AggregateNetworkData.cpp:217]: Fine - OK!

File: src/opm/output/eclipse/AggregateNetworkData.cpp (hunk 6 above)
Comment (jalvestad) [reply src/opm/output/eclipse/AggregateNetworkData.cpp:113]: No special reason! - fixed!

File: src/opm/output/eclipse/AggregateNetworkData.cpp (hunk 7 above)
Comment (jalvestad) [reply src/opm/output/eclipse/AggregateNetworkData.cpp:115]: Fine -OK, changed.

File: src/opm/output/eclipse/AggregateNetworkData.cpp (hunk 8 above)
Comment (jalvestad) [reply src/opm/output/eclipse/AggregateNetworkData.cpp:133]: yes- agree, changed.

File: src/opm/output/eclipse/AggregateNetworkData.cpp (hunk 9 above)
Comment (jalvestad) [reply src/opm/output/eclipse/AggregateNetworkData.cpp:246]: The variable node_name may in some cases change to the name of an uptree node with fixed pressure. The nodeName is a const.

File: src/opm/output/eclipse/AggregateNetworkData.cpp (hunk 10 above)
Comment (jalvestad) [reply src/opm/output/eclipse/AggregateNetworkData.cpp:258]: OK - done. (Note: I used a copy of code used in other parts of opm-common - are all throw statements to be rewritten? - now there are different implementations in use)

File: src/opm/output/eclipse/AggregateNetworkData.cpp (hunk 11 above)
Comment (jalvestad) [reply src/opm/output/eclipse/AggregateNetworkData.cpp:310]: OK - fine - done.

File: src/opm/output/eclipse/AggregateNetworkData.cpp (hunk 12 above)
Comment (jalvestad) [reply src/opm/output/eclipse/AggregateNetworkData.cpp:430]: OK - done.

File: src/opm/output/eclipse/AggregateNetworkData.cpp (hunk 10 above)
Comment (joakim-hove) [reply src/opm/output/eclipse/AggregateNetworkData.cpp:258]: > OK - done

Good!

>are all throw statements to be rewritten? 

To rewrite existing statements is maybe a bit over the top - but new code should use fmtlib. Personally I also update old code to use fmtlib "when I am there".

File: src/opm/output/eclipse/AggregateNetworkData.cpp (hunk 13 above)
Comment (jalvestad) [reply src/opm/output/eclipse/AggregateNetworkData.cpp:472]: The meaning of the value of this index is unknown, but it has to be 1!

File: src/opm/output/eclipse/AggregateNetworkData.cpp (hunk 15 above)
Comment (jalvestad) [reply src/opm/output/eclipse/AggregateNetworkData.cpp:604]: The meaning of the value of this index is unknown, but it has to be 1.!

File: src/opm/output/eclipse/AggregateNetworkData.cpp (hunk 13 above)
Comment (joakim-hove) [reply src/opm/output/eclipse/AggregateNetworkData.cpp:472]: Ok, add a comment?

File: src/opm/output/eclipse/AggregateNetworkData.cpp (hunk 14 above)
Comment (jalvestad) [reply src/opm/output/eclipse/AggregateNetworkData.cpp:595]: OK - fine - done.

File: src/opm/output/eclipse/AggregateNetworkData.cpp (hunk 16 above)
Comment (jalvestad) [reply src/opm/output/eclipse/AggregateNetworkData.cpp:718]: OK - rewritten.

File: src/opm/output/eclipse/AggregateNetworkData.cpp (hunk 13 above)
Comment (jalvestad) [reply src/opm/output/eclipse/AggregateNetworkData.cpp:472]: OK - I will write comments for these special cases.

**Comment by jalvestad** [comment]
All content in this PR (apart from a very minor correction) is needed to output the restart-data for the extended Network option. As I see it, it is an advantage that all changes for this purpose are included, so that it is possible to understand why the changes made, that they are needed for a purpose. Bård Skaflestad (bska) has been kept informed and given me advice during the development of the PR - also being a maintainer for opm-common and I have the understanding that he has not objected/disagreed in making this as one pull request. All in all I prefer to keep this as one PR for this case, although understanding you do not consider this as the best solution. 

I will soon update the PR with the changes made, however, there is an issue using the recommended function sumState.get_group_var(nodeName, "GPR", 0.) which I need to solve for the code to work correctly.

**Comment by joakim-hove** [comment]
>All in all I prefer to keep this as one PR for this case, although understanding you do not consider this as the best solution.

Sure - my main point is that I would urge you to create small PR's as you go when things are ready.

>there is an issue using the recommended function sumState.get_group_var(nodeName, "GPR", 0.)


Ok - I was a bit uncertain about that myself, maybe it was bad advice.

**Comment by jalvestad** [comment]
All content in this PR (apart from a very minor correction) is needed to output the restart-data for the extended Network option. As I see it, it is an advantage that all changes for this purpose are included, so that it is possible to understand why the changes made, that they are needed for a purpose. Bård Skaflestad (bska) has been kept informed and given me advice during the development of the PR - also being a maintainer for opm-common and I have the understanding that he has not objected/disagreed in making this as one pull request. All in all I prefer to keep this as one PR for this case, although understanding you do not consider this as the best solution. 

I will soon update the PR with the changes made, however, there is an issue using the recommended function sumState.get_group_var(nodeName, "GPR", 0.) which I need to solve for the code to work correctly. 

The latest version is now tested and found OK with the changes made, and is ready for merge from my side. 

Thanks for good advice for improving the code and resolving issue with  get_group_var() function!

**Comment by joakim-hove** [comment]
jenkins build this update_data please

**Comment by bska** [comment]
> update_data

Please note that having multiple data updates in flight (this and OPM/opm-simulators#3067, OPM/opm-tests#481) almost always leads to conflicts.

**Comment by bska** [comment]
jenkins build this opm-test=482 please

**Comment by bska** [comment]
jenkins build this opm-tests=482 please

**Comment by bska** [comment]
New reference solutions have been merged to master.  I'll merge this now to bring the system back to a consistent state.
